server/src/main/kotlin/tech/figure/objectstore/gateway/components/BlockStreamConsumer.kt

```diff
@@ -0,0 +1,103 @@
+package tech.figure.objectstore.gateway.components
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import mu.KLogging
+import org.springframework.beans.factory.annotation.Qualifier
+import org.springframework.boot.actuate.health.Health
+import org.springframework.boot.actuate.health.HealthIndicator
+import org.springframework.boot.context.event.ApplicationReadyEvent
+import org.springframework.context.ApplicationListener
+import org.springframework.stereotype.Component
+import tech.figure.objectstore.gateway.configuration.BeanQualifiers
+import tech.figure.objectstore.gateway.configuration.BlockStreamProperties
+import tech.figure.objectstore.gateway.eventstream.GatewayEventStream
+import tech.figure.objectstore.gateway.repository.BlockHeightRepository
+import tech.figure.objectstore.gateway.service.StreamEventHandlerService
+import java.util.concurrent.atomic.AtomicBoolean
+import kotlin.time.Duration.Companion.seconds
+import kotlin.time.ExperimentalTime
+
+@Component
+class BlockStreamConsumer(
+    private val streamEventHandlerService: StreamEventHandlerService,
+    private val blockHeightRepository: BlockHeightRepository,
+    private val blockStreamProperties: BlockStreamProperties,
+    private val gatewayEventStream: GatewayEventStream,
+    @Qualifier(BeanQualifiers.BLOCK_STREAM_COROUTINE_SCOPE_QUALIFIER) private val blockStreamScope: CoroutineScope,
+) : ApplicationListener<ApplicationReadyEvent>, HealthIndicator {
+    private companion object : KLogging() {
+        private var blockStreamRunning = AtomicBoolean(false)
+        private val TX_EVENTS = setOf("wasm")
+    }
+
+    override fun onApplicationEvent(event: ApplicationReadyEvent) {
+        if (!blockStreamProperties.enabled) {
+            logger.warn("Event stream has been manually disabled! Use a value of `true` for EVENT_STREAM_ENABLED to re-enable it")
+            return
+        }
+        tryStartBlockStream {
+            blockStreamLoop()
+        }
+    }
+
+    override fun health(): Health = if (blockStreamRunning.get()) {
+        Health.up().build()
+    } else {
+        Health.down().build()
+    }
+
+    private suspend fun blockStreamLoop() {
+        val lastBlockProcessed = blockHeightRepository.getLastProcessedBlockHeight()
```

Comment:
This is really cool how you can just mod your service on the fly with a redeploy to switch between the ES and Blockapi.  <3